### DISCUSSION TITLE
Ebusd Settings

### DISCUSSION THREAD

**Discussion by RotorMotor42** [cyberthom42/vaillant-arotherm-plus#6] [General]
Ich hole die Daten per MQTT, habe das ebusd pollinterval auf 1 Sekunde und schreibe die Daten in eine InfluxDB.
Wie macht ihr das?

**Comment by cyberthom42** [comment]
Ich nutze den ebus 2 auf einem Raspberry. Ich nutze iobroker mit ebusd Plugin (Verbindung über Port 8888/8889). Hier kann man automatisch die Objekte in eine InfluxDB schreiben. Ich will dann auch über Scripte benutzerdefinierte Werte schreiben, wie z.B. realer Stromverbrauch WW und realer Stromverbrauch Heizung, realer Stromverbrauch Standby. Das Script hatte ich angefangen, aber noch nicht fertig,

**Comment by cyberthom42** [comment]
Interessant wäre zu wissen, ob ein Interval von 1 Abfrage / Sekunde irgendwie schädlich für die Steuerung ist bzw. ob es da eine Beeinträchtigung der Steuerung geben kann. 
Kannst du prüfen, ob die Abfragen wirklich an den Bus gesendet werden (also mit Option "-f ") oder an den Cache.  Ich glaube den Cache Timer kann man auch einstellen.

**Reply by RotorMotor42** [reply]
Ich nutze die MQTT Schnittstelle um mich auf Werte zu subscriben.

Durch das ?1 wird der Wert in jedem Cycle vom Bus angefragt und dann über MQTT bereitgestellt:
`mosquitto_pub -m '?1' -t 'ebusd/hmu/HE_EEV_Auslass/get'`

In Node Red habe ich dann nur noch einen mqtt listener, der dann die werte auf ebusd/+ entgegen nimmt.

So erhalte ich jede Sekunde einen Wert, aber da es viele Werte sind, auf die ich mit ?1 oder ?2 subscribe habe ich pro Wert nur ca. 0.03Hz.

**Comment by cyberthom42** [comment]
Die Vorlauftemperatur tracke ich mit HE_Vorlauftemperatur. Ich habe festgestellt, dass FlowTemp stark gerundet wird.

**Reply by RotorMotor42** [reply]
für mich ist 
`hmu FlowTemp` der genauste Wert mit zwei Nachkommastellen.
`hmu HE_Vorlauftemperatur` hat bei mir nur eine.
`ctlv2 Hc1FlowTemp` ist am ungenausten mit einem halben Grad Genauigkeit

Allerdings wäre es auch möglich, dass an unterschiedlichen Stellen gemessen wird?
Und die Frage ist, welcher dieser Werte geht ins Energieintegral ein?

**Reply by cyberthom42** [reply]
Ah stimmt, da gibt es ja zwei FLowTemp, ich meinte den ctlv2/basv Wert vom Controller. Der ist ungenau. Interessant wäre der Unterschied zwischen `hmu FlowTemp` und `hmu HE_Vorlauftemperatur`. Die arotherm hat ja nur einen Sensor, der ist meine ich draussen. Die Hydraulikstation hat glaube ich keinen weiteren Sensor.

**Comment by RotorMotor42** [comment]
@cyberthom42 
Funktionieren bei dir eigentlich die "Total" Werte?
Bei mir liefern sie: hmu ConsumptionTotal =  (ERR: invalid position for 3308b516021802 / 00)
hmu YieldTotal =  (ERR: invalid position for 3308b516021801 / 00)

YieldHc und Hwc scheinen allerdings zu funktionieren.
Wie wertest du eigentlich solch steigende Werte aus?

**Reply by cyberthom42** [reply]
Die hmu Total Werte gehen bei mir auch nicht, aber basv (bei dir ctlv2) YieldTotal geht bei mir. Ich muss zugeben, das Thema logging von Energiewerten habe ich noch nicht final und zufriedenstellend laufen aus Zeitmangel.
Aktuell logge ich möglichst alles, Auswertung kann ich dann später machen. 
Für Energie habe ich aktuell folgende Auswertung, die gibt mir auch ein ganz gutes Bild. Dafür nutze ich YieldHcDay/YieldHwcDay und meine gemessenen Energiewerte. Ich hatte auch ein Script angefangen, welches mir auch die unterschiedlichen Energieverbräuche für Heizen, ww, standby, enteisen aufschlüsselt, hatte ich noch nicht fertig gestellt. 
![image](https://user-images.githubusercontent.com/115342533/209718871-ab1aa343-006f-45a3-8398-cf3c7070da32.png)

**Reply by cyberthom42** [reply]
> @cyberthom42 Funktionieren bei dir eigentlich die "Total" Werte? Bei mir liefern sie: hmu ConsumptionTotal = (ERR: invalid position for 3308b516021802 / 00) hmu YieldTotal = (ERR: invalid position for 3308b516021801 / 00)
> 
> YieldHc und Hwc scheinen allerdings zu funktionieren. Wie wertest du eigentlich solch steigende Werte aus?

In Grafana nutze ich "spread", das gibt für den Zeitraum die Differenz von kleinsten zum größten Wert, also genau das was man hier benötigt. Die "Day" Werte von Vaillant sind etwas Tricky, da sie nicht 0h wieder auf "0" gesetzt werden sondern ein paar Minuten später. Bei den Day Werten nutze ich daher immer "last"

**Comment by chrizzzp** [comment]
Nochmal eine Frage zum ebusd-Pollintervall/Frequenz des Datenauslesens:
Welche Methode hat sich eigentlich bewährt für eine zuverlässige und zeitlich möglich hochaufgelöste Abfrage der ebus-Daten?
Ich habe z.B. in meiner `08.hmu.csv` für einige Werte die Pollfrequenz auf maximal gesetzt (r1), habe aber das Gefühl, daß die Daten doch nur sporadisch (alle paar Minuten) per MQTT ankommen (lt. MQTT-Explorer). 
z.B.
`r1,,CompressorUtilization,,,,,25,,,UIN,10,%,,,,,,,,,,,,,,,,,,,`
`r1,,EnergyIntegral,,,,,21,,,integral,,°min,current energy integral in °min,,,,,,,,,,,,,,,,,,`

Pollintervall von ebus ist derzeit auf 5s (default) gesetzt. Wenn ich dagegen z.B. mit 
`ebusctl read -f -c hmu CompressorUtilization` 
ein Auslesen erzwinge, dann kommen die Werte sofort per MQTT an (geht auch ohne -f, dann eben aus dem Cache). 

Kann mir das nicht wirklich erklären.

Die wichtigen Werte also doch lieber "manuell" per ebusctl-Skript abfragen (ggf. über 'cron')?

Bei mir läuft ebusd übrigens im eigenen Docker-Container (ebusd-Integration von Home Assistant).

**Reply by cyberthom42** [reply]
Ich habe MQTT für mich auch erst einmal aufgegeben, da ich nicht zufrieden mit der Implementierung bin.
Das Setting (r1) ist nicht die Polling-Frequenz, das ist die Priorität. Lies dir mal die Beschreibung des ebus Wiki dazu durch. Ich versuch es mal zu erklären wobei bei mir da noch Detailfragen offen sind, die nirgends beantwortet sind. 
Nach meinem Verständnis werden die Abfragen im PollingInterval (also hier 5s) durchlaufen. Allerdings sequentiell. Wenn du z.B. 5 Werte mit r1 hast, wird dann als erstes der erste Wert abgefragt, nach 5 Sekunden der nächste usw. die Werte r2 und r3 usw. werden dann nicht mit jedem Zyklus abgefragt. Es gibt hierzu ein Issue bzw. Discussion auf dem ebusd github. Wenn ich das finde poste ich mal den Link hier. 
Ich nutze sowieso iobroker da ich HA nicht mag. Dort gibt es ein ebus Plugin, das entspricht in etwas einer Abfrage per Scipt. Hiermit erhalte ich wesentlich genauerer Ergebnisse. Das Problem ist ja, dass viele Werte eine Zeiltliche Korrelation haben, da macht es wenig Sinn, wenn die Abfrage 20 Sekunden Differenz hat. Man muß allerdings beachten, dass man nicht einfach 50 Abfragen auf einmal abschicken kann, das könnte zu einer Überforderung führen bzw. werden diese dann so schnell wie möglich abgearbeitet. Ich frage sehr viele Werte auf einmal ab, allerdings dann aller 30 Sekunden.

**Reply by cyberthom42** [reply]
Einzelne Parameter frage ich mit einem separaten Script (bzw. weiterer Instanz) öfter ab. Manche dann auch aller 10 Minuten

**Reply by chrizzzp** [reply]
Habe mich etwas missverständlich ausgedrückt. Ich habe soweit verstanden, daß 'r1' eine Prioritätsangabe ist (die höchste von 1-9, d.h. mit r1 erreiche ich die maximal mögliche Pollfrequenz). Du meintest vielleicht diesen Thread in dem John das mal erklärt hat: https://github.com/john30/ebusd/issues/458

So wie ich das nach seiner Erklärung verstanden hatte, werden alle Poll-Requests mit Priorität 1 in jedem Poll-Zyklus abgefragt. Aber wenn ich mich recht erinnere, kann es manchmal sein, daß der Bus nicht "frei" ist und dann bleiben die Requests vermutlich länger (ggf. mehrere Poll-Intervalle) in der Queue (Stichwort: Arbitrierung?).

Also Dein Pollintervall ist 30 Sekunden? Kommen denn alle Werte regelmässig an, wenn man nur alle 30 Sekunden abfragt?

Ich kenne bisher nur HA. Habe aber gehört, es ist ein bissel Religion, ob man iobroker, HA etc. mag... ;-)

**Reply by cyberthom42** [reply]
Bei mir kommen alle Werte an, auch die bei denen ich aller 10s abfrage. Man darf es eben nicht übertreiben, dann gehen Werte verloren bzw. überschneiden sich. 
Meine oben geschilderte Vermutung basiert auf die Aussage von John, du hast den richtigen Beitrag verlinkt:
"the idea behind it is that the bus is penetrated only once every pollinterval seconds. so every pollinterval seconds another message gets picked from the priority queue of messages to be polled. a poll cycle would be complete after cycling through all messages having the same lowest priority (usually 1)."

"every pollinterval seconds another message gets picked from the priority queue of messages to be polled"
=> hier verstehe ich, dass jeweils immer nur eine Message abgefragt wird

" a poll cycle would be complete after cycling through all messages having the same lowest priority (usually 1).""
=> hier verstehe ich, dass ein cycle = n (Anzahl r1 Einträge) * Polling Interval (bei dir 5s).

Wann kommen deine Nachrichten an? Du müsstest doch sehen, ob die alle zur gleichen Zeit ankommen oder mit einem Versatz von 5s

**Reply by cyberthom42** [reply]
Du kannst das ja mal mit `mosquitto_sub -h xxx.xxx.xxx.xxx -t "ebusd/#" -F '@H:@M:@S %t %p' `prüfen, da siehst du den Zeitsptempel. Topic evtl. anpassen

**Reply by cyberthom42** [reply]
Ich hab mal bei mir Debug eingeschaltet. Mein iobroker Script fragt 80 Werte ab. Mein Abfrageintervall ist 60 Sekunden (nicht 30 Sekunden, da hatte ich mich vertan). Die Abfrage der 80 Werte dauert ca. 40 Sekunden. Also dauert eine Abfrage im Schnitt ca. 0,5 Sekunden. Daher bezweifle ich, dass alle r1 Werte innerhalb des Intervals von 5s abgefragt werden, das schafft der Bus nicht bzw. dürften es nur max. 10 Werte sein.

**Reply by chrizzzp** [reply]
Ich habe zur Zeit ca. 15 Werte in CSVs auf Priorität 'r1' gesetzt (Rest ist unverändert 'r'). Per Skript (ebusctl read -f...) frage ich derzeit drei der 15 Werte alle 60s ("Abfrageintervall") ab. ebusd pollt alle 5s ("Pollintervall").

Alle drei per Skript abgefragten Werte kommen meist zur gleichen Sekunde per MQTT an, ca. alle 60 Sekunden. Daher vermute ich, daß in einem Pollzyklus mehrere Messages gesendet und empfangen werden können. Das bestätigt Deine Messung m.E. auch. 

Da der Ebus ja recht langsam ist (2400 baud) kann es vielleicht vorkommen, daß z.B. nicht alle 'r1'-Werte bis zum nächsten Poll (+5s) übermittelt sind. Dann könnten die Zeitabstände/Pollzyklen aus dem Takt kommen. Das sehe ich sporadisch auch. Bei mir ist auch ohne meine Abfragen eine ganze Menge Traffic auf dem Ebus (insbesondere durch die Sensoren-Eingänge vom VR71 Mischermodul). HA hat per MQTT-Autodiscovery über 220 Entitäten von ebusd angelegt. Die Werte davon kommen aber bei weitem nicht alle regelmässig über den Ebus.

Wie sieht Dein "Abfrageintervall" (Skript) im Vergleich zum "Pollintervall" (ebusd) aus? Ich vermute mal hier muß man je nach "Traffic" das richtige Verhältnis finden.

**Comment by mf0o** [comment]
Ist zwar Offtopic aber Ich finde die Diskussion hier super interessant.  
Wenn erlaubt, würde Ich mich mit einklinken. Ich muss meine Arbeiten nun auch endlich wieder aufnehmen (Vaillant Flexotherm mit Arocollect und eBUS Adapter v5) um endlich meine Wärmepumpe nach Überschuss der PV steuern zu können.

**Reply by chrizzzp** [reply]
Genau das ist auch mein Ziel, zusätzlich zum (möglichen) Eingriff in die Vaillant-Regelung (Stichwort: Takten überdimensionierter Wärmepumpen... ;-)). Mit der besseren Verfügbarkeit der eBUS Adapter (z.B. V5 von John) mehrt sich vielleicht auch nach und nach die Zahl der Interessenten?

**Reply by cyberthom42** [reply]
Habt ihr schon den Bereich im Wiki dazu angesehen? PV-Überschuss Steuerung ist eigentlich möglich. Man kann es jedoch so steuern, dass die potentiell "guten" PV-Zeiten besser von der WP genutzt werden. Aber die WP kann nicht so schnell auf PV-Überschuss Änderungen reagieren wie z.B. eine Wallbox. Ich lege bei mir die WP in der Übergangszeit in die Zeit in der ich eigentlich immer genug Strom habe und hebe hier dann aber die Solltemp an. Was man jedoch machen kann wäre eine intelligentere Steuerung die die Vorhersagen verwendet. Ich monitore bei mir die Vorhersage von 2 Anbietern. Aber auch hier sieht man schon dass sich selbst die Vorhersagen innerhalb eines Tages von Stunde zu Stunde z.T. stark ändern. Die Tendenz passt aber of.

**Reply by cyberthom42** [reply]
Hier mal ein Beispiel, habe solcast und forcast.solar
![image](https://github.com/cyberthom42/vaillant-arotherm-plus/assets/115342533/4893abb8-33e2-4219-b69d-487e4c538d40)

**Reply by mf0o** [reply]
Richtig gut @cyberthom42! Die Wiki
Page hatte ich gesehen, konnte aber vor Projektstart nicht viel damit anfangen. 

In erster Linie möchte ich die „Funktion“ *hust* der SG ready Schnittstelle nachbauen (Solltemperatur bei Überschuss erhöhen). 
Im zweiten Schritt die Schalt- und Heizphasen monitoren um ggf. die Konfiguration zu optimieren. 
Aus der API von Solaredge bekomme ich schon alles notwendige von der PV. Beim eBUS habe allerdings bisher nur die hardware am laufen und dank autoscan und CLI Abfrage ein paar Werte gesehen.
 Zu mehr bin ich mangels Zeit noch nicht gekommen. 

Dank euren Antworten hier erscheint mir das ganze wieder eher als machbar :)

**Reply by chrizzzp** [reply]
@cyberthom42 : Welche Temperatur hebst Du bei PV-Überschuss an und wie (konstanter Offset oder abhängig vom Überschuss?). Ich vermute Du benutzte dann nicht den ME-Eingang sondern steuerst direkt über Ebus?
Bei mir gibt es `hmu SetMode flowtempdesired` und `ctlv2 Hc1ActualFlowTempDesired`. Da ich keinen Direktkreis habe (Heizkreis 1 hinter Mischer) macht es einen Unterschied. Erhöhung von `hmu SetMode flowtempdesired` würde m.E. meinen parallelen Pufferspeicher aufheizen, Erhöhung von `ctlv2 Hc1ActualFlowTempDesired` würde den Estrich der FBH (und den Pufferspeicher) aufheizen.

**Reply by cyberthom42** [reply]
Ich muss die Euphorie hier wohl ein wenig bremsen. Nach meiner Erfahrung ist eine PV-Überschuss Steuerung eigentlich nicht sinnvoll möglich, schon gar nicht mit dem SG-Ready Eingang. Gerade wenn Sonne und Wolken wechseln und dabei noch gekocht wird kann es ja zu sehr starker Schwankung beim PV-Überschuss kommen. Damit kommt die Arotherm nicht klar.  Dann fällt das Integral relativ schnell ab und die WP schaltet ab. Das kostet Effektivität und bringt eigentlich das Gegenteil. Was gut funktioniert, ist die normalerweise "guten" Zeiten für PV-Überschuss zu nutzen. So mache ich es. Gut wäre dann noch eine Wetter/Solar Prognose mit einzubinden. 
Zur eigentlichen Frage: Es gibt mehrere Möglichkeiten. Ich hebe die Min Soll-VL Temp an.

**Reply by chrizzzp** [reply]
Ja, da mache ich mir auch keine allzu großen Illusionen. Trotzdem würde ich erwarten, daß es in der Übergangszeit trotzdem vorteilhafter ist, die WP auch bei schwankendem PV-Überschuss laufen zu lassen - sofern man die Wärme speichern kann - als ausschließlich nach AT-Regelung (oft ganz ohne PV-Unterstützung, z.B. nachts) - auch wenn man dabei den Netzbezug nicht ganz vermeiden kann. Aber klar, Takten durch schwankenden PV-Überschuss sollte man vermeiden.

Die Prognosen zu berücksichtigen ist ein interessanter Ansatz.  Ich könnte mir vorstellen, zusätzlich einen zeitlich gemittelten Wert des PV-Überschusses heranzuziehen, um zu entscheiden, ob die WP aktiviert wird.

Was meinst Du mit "normalerweise "guten" Zeiten"?
